[SYCL] diagnose any pointers captured into sycl kernel

cperkinsintel opened the pull request:
Signed-off-by: Chris Perkins <chris.perkins@intel.com>

 Any pointer captured into the SYCL kernel lambda will fail when dereferenced...except USM. If it weren't for USM we could just emit a diagnostic for every pointer capture. Instead, we attempt to identify which pointers are USM, and which are definitely not (and will crash). For those that will crash, we emit an error. For those that are unknown, we _could_ emit a gentle note suggesting the user double check that they are using USM.  But at this time we do not. 
For safe USM pointers, we do nothing.

The analysis of the capture must be done after the parent function has an intact body, but the Sema call to check captures comes before that, when the lambda is being built.  So we have two calls: one that records any captures that need checking, and another that performs the diagnostics when the parent function is being finished.

cperkinsintel commented:
> We should be able to just act on the lambda capture event, diagnose whether it is a USM THERE, then use the deferred diagnostics to see if this is called.

That might work, I'm not familiar enough with the deferred diagnostic system to say. With our existing deferred diagnostics, the condition is fairly simple: is this code going to be in a kernel or not? And if so, then we emit anything outstanding.    Doing what you suggest seems like it would introduce a whole new class of deferred diagnostics. I can investigate.

>Also, this fails pretty horribly in the event of reference captures that are modified after the lambda is created, but not yet called.
No, this case is handled, and there are test cases for what you describe.  There is no assumption on a particular order of execution.   This users code has a pointer reference whose initialization has been assessed to be dangerous. Then we further examine the parent context to see if there are any other references to it, ie maybe it was passed as an argument to another function, or maybe it was re-assigned later. If there is anything at all, then we back away from emitting an error - we trust the user knows what they are doing.   And we don't worry about order, too hard to figure out.

cperkinsintel commented:
@erichkeane   the deferred diagnostics track in-context declarations and then visit them later. I tested it and I can switch to use that. So there will be just one function to check and diagnose illegal captures, and no multimaps or anything to store and retrieve captures.  So that's a definite improvement.

I also looked into AnalysisBasedWarnings, but I don't see anything general there that would be usable for what needed here. There is definitely a lot of lifetime analysis going on there, for finding unreachable code, exceptions in no-throw functions, etc. but it's all quite specific. 

I'll finish up the change to leverage the deferred diagnostics better.  Let me know if you have any other guidance or requests.

cperkinsintel commented:
Also, anyone have a clue on what is going on with the Windows buildbot?  I tried it on a local windows machine and it does not have the same error.

erichkeane commented on clang/lib/Sema/Sema.cpp:
```diff
@@ -45,6 +45,9 @@
 #include "llvm/ADT/SmallSet.h"
 #include "llvm/Support/TimeProfiler.h"
 
+//CP
```
Remove please.

erichkeane commented on clang/lib/Sema/Sema.cpp:
```diff
@@ -1520,7 +1523,7 @@ class DeferredDiagnosticsEmitter
     --InOMPDeviceContext;
   }
-
+  
```
Unrelatedchange.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -283,6 +283,178 @@ void Sema::checkSYCLDeviceVarDecl(VarDecl *Var) {
   checkSYCLVarType(*this, Ty, Loc, Visited);
 }
 
+bool isVarInExpr(VarDecl *V, std::function<bool(const Expr *)> AF,
```
Free functions NEED to be static.

erichkeane replied:
Also, don't use std::function here.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -283,6 +283,178 @@ void Sema::checkSYCLDeviceVarDecl(VarDecl *Var) {
+  ExprAllocation howAllocated = Unknown;
+
+  // We will use this to check declarations and assignments.
+  auto VetteCallExpr = [&howAllocated](const Expr *E) {
```
it is spelled 'Vet'.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -283,6 +283,178 @@ void Sema::checkSYCLDeviceVarDecl(VarDecl *Var) {
+
+  // Emit diagnostics.
+  if (howAllocated == WillCrash) {
+    Diag(CaptureLoc, diag::err_sycl_illegal_memory_reference);
```
I thought you were switching these to delayed diagnostics?  This using them.

cperkinsintel replied:
It's now called _by_ the delayed diagnostic system. So we emit a normal one.

erichkeane replied:
That is inverted, thats not how this should work.

cperkinsintel replied:
I ended up reverting and rebasing. Am emitting deferred diagnostics again. But see my standalone comment about this

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -283,6 +283,178 @@ void Sema::checkSYCLDeviceVarDecl(VarDecl *Var) {
+        howAllocated = WillCrash;
+
+      const UnaryOperator *UO = dyn_cast<UnaryOperator>(Init);
+      if (UO && (UO->getOpcode() == UO_AddrOf))
```
I'm guessing this logic is going to be quite a bit of false errors.  You cannot really do that.

cperkinsintel replied:
This is only screening variable initialization (of pointers we want to capture in the kernel lambda), it's not forbidding the use of the address_of operator generally.
 If a pointer variable was initialized by taking the address of another variable, that's not going to be capturable. That will segfault.     So, once the other analysis is done and we are confident the variable wasn't otherwise referenced, we can emit the error in confidence.  Am I overlooking an obvious counter-example?

cperkinsintel replied:
you are correct. Removed.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -283,6 +283,178 @@ void Sema::checkSYCLDeviceVarDecl(VarDecl *Var) {
+  // is subject to change. So we look through the statements to see if that
+  // variable is otherwise referenced before capture, or if reassigned it is
+  // assessed with AssignF.
+  DeclContext *DC = Var->getParentFunctionOrMethod();
```
This part is really bad too.  What you  SHOULD be doing is walking backward on uses in the statement and see where it was last set, not attempt to derive it by walking the function.

cperkinsintel replied:
You mean with a DeclRefVisitor? I'm not sure how else to walk the uses.

erichkeane replied:
Check out AnalysisBasedWarnings' Wuninitialized handling.  It has infrastructure to do exactly what you want.

cperkinsintel replied:
will do

cperkinsintel replied:
done

erichkeane commented:
Check out the -Wuninitialized code in AnalysisBasedWarnings, it is almost exactly what you want, except the only 'use' you care about is capturing, and instead of 'init', you care about how it originated.  This current patch has a convoluted and error-prone mechanism to attempt to detect this, while there is existing infrastructure to do very nearly this.

cperkinsintel commented:
@erichkeane  - I looked at -Wunitialized ContainsReference as you suggested and made a new analyzer (HasRefToVar) modelled on it.  That's working well. I also have the code emitting deferred diagnostics instead direct ones, as requested. 

But I had to revert back to using separate functions for recording captures and diagnosing. The issue is that at the time the capture is being identified in the lambda, it may be complete, but its parent function, where we need to search for other references to the captured declaration, might not have a body at that time.  So I record the potential captures and the lambda FunctionDecl at that time, but keyed off the parent FunctionDecl. Then when the parent function is finally complete, we finish the analysis and emit deferred diagnostics. 
I don't like having two functions for this, and it seems like its reproducing a bit of what the deferred diagnostic system is supposed to do itself. But the problem is those two different function decls. One for the parent, that needs to be searched, and one for the lambda (which is what the SYCL kernel deferred diagnostics are keyed to).    I looked into sub-classing or otherwise extending the deferred diagnostic system, but it looks like I'd have to change how those diagnostics are called, and that would touch a lot of code , including every DD consumer (OpenMP, CUDA, etc).   

Let me know your thoughts.

erichkeane commented:
> 
> 
> @erichkeane - I looked at -Wunitialized ContainsReference as you suggested and made a new analyzer (HasRefToVar) modelled on it. That's working well. I also have the code emitting deferred diagnostics instead direct ones, as requested.
> 
> But I had to revert back to using separate functions for recording captures and diagnosing. The issue is that at the time the capture is being identified in the lambda, it may be complete, but its parent function, where we need to search for other references to the captured declaration, might not have a body at that time. So I record the potential captures and the lambda FunctionDecl at that time, but keyed off the parent FunctionDecl. Then when the parent function is finally complete, we finish the analysis and emit deferred diagnostics.
> I don't like having two functions for this, and it seems like its reproducing a bit of what the deferred diagnostic system is supposed to do itself. But the problem is those two different function decls. One for the parent, that needs to be searched, and one for the lambda (which is what the SYCL kernel deferred diagnostics are keyed to). I looked into sub-classing or otherwise extending the deferred diagnostic system, but it looks like I'd have to change how those diagnostics are called, and that would touch a lot of code , including every DD consumer (OpenMP, CUDA, etc).
> 
> Let me know your thoughts.

I don't think this even works... If you wait until the end of the function, the value of some captures would have changed since they were called.  Consider:

`
void foo() {
  int *Ptr = malloc_shared(...);
  auto lambda = [&Ptr](){};

  lambda(); // Should be fine here.
  Ptr = malloc(...);
  lambda(); // Should NOT be fine here.
 }  
`

If you wait to the end of the function, you'll not end up diagnosing the right values, right?  You can't start at the lambda creation location, you need to start at the lambda CALL location, then step backwards from there.  I think the only sensible place to diagnose this 

ALSO, the statements for the function "So far" actually already exist.  You shouldn't have to do the 2 functions, you just have to get the function statements.

cperkinsintel commented:
@erichkeane  - Look at the lit test I added ( sycl-pointer-capture-diagnostics.cpp ) to get a better sense for it. I just pushed some more "unknown" tests that make it even more explicit. Look at not just what captures are tripping errors, but which ones do not.    I just added some more variables like `goodThenBad, badThenGood, goodThenBadThenGood`  to make this even more explicit, and  for all of those it does NOT emit an error.   But that's not the same as the code not working. The code works by precisely emitting errors when the situation is reasonably determinable, and not emitting anything when the situation is at all questionable.   In the questionable situation ,we have to trust that the developer knows what they are doing. 

In the example you provide, there is no way to safely assert that Ptr is good or that it is bad if captured by the kernel lambda.   And this code, correctly, will NOT emit an error in this case. It sees those other references and correctly determines that it CANNOT safely emit an error and we must trust that the user knows what they are doing.  

It only emits errors for variable capture when it the situation is assured to be problematic. Fortunately, the common cases aren't complex and are easily caught. 

> but that only works if the lambda was created and called in the same function. Else, it wouldn't work.  

The lambda can be called anywhere. That doesn't matter. 


>  I'm beginning to think this isn't a particularly good error for us to implement. I think it belongs in a static analyzer

FWIW, CodePlay throws an error for any pointer variable captured by a kernel lambda. But they dont' have to worry about USM support.   To my mind, since we are trying to provide users with both USM and a compiler it seems we should do what we can to help them.

erichkeane commented:
If we are going to diagnose something in the compiler, it needs to catch basically every case. That is just how good compiler diagnostics work. THIS is not a good compiler diagnostic and I don't believe it can be without solving the halting problem.

So I am against accepting this diagnostic unless it can be as reliable as compiler diags are supposed to be.

erichkeane commented:
Also, I haven't spent much time on your test because it is quite disorganized and isn't really possible to follow what is going on. This test needs a massive amount of improvement to be readable

cperkinsintel commented:
>  I don't believe it can be without solving the halting problem

Yes, it is as good as it can be without taking on the halting problem.  ( * )

> So I am against accepting this diagnostic unless it can be as reliable as compiler diags are supposed to be.

A warning then? You and @Fznamznon  are the Code Owners and the decision is yours, but a message in this domain has value.  Remember, _any_ pointer captured by the kernel lambda is going to segfault when dereferenced ... except USM.  So even if we can't catch the complex cases, having something for the simpler ones can really save some developer from a lot of pain.   

>  This test needs a massive amount of improvement to be readable.

If you have suggestions or requests, I'll be happy to make the changes. As it is, I know there are a lot of cases, but I tried to sort out the declarations with comments between sections:  crashing pointers that are caught, usm, malloc, structs, pointers that can't be caught.  And then repeat those sections and variables in the same order (with the same comments) down below where the capture occurs.  The test opens with the normal mock sycl namespace and mock USM functions, and, while labelled, maybe I could make it clearer to scroll down to get into the actual tests. Or maybe I should move those mocks to an #included file.  



( * ) There's an old joke about this, goes something like:
Q: How long will it take you to solve the halting problem?
A: Let me do some calculations and get back to you....

erichkeane commented:
> > I don't believe it can be without solving the halting problem
> 
> Yes, it is as good as it can be without taking on the halting problem. ( * )
> 
> > So I am against accepting this diagnostic unless it can be as reliable as compiler diags are supposed to be.
> 
> A warning then? You and @Fznamznon are the Code Owners and the decision is yours, but a message in this domain has value. Remember, _any_ pointer captured by the kernel lambda is going to segfault when dereferenced ... except USM. So even if we can't catch the complex cases, having something for the simpler ones can really save some developer from a lot of pain.

No, not any sort of diagnostic in the compiler.  Users depend on compiler diagnostics to catch their mistakes, and when you have one as so easy to trick as this one they cannot depend on them. Clang's community opinion on these is to simply reject them.

> 
> > This test needs a massive amount of improvement to be readable.
> 
> If you have suggestions or requests, I'll be happy to make the changes. As it is, I know there are a lot of cases, but I tried to sort out the declarations with comments between sections: crashing pointers that are caught, usm, malloc, structs, pointers that can't be caught. And then repeat those sections and variables in the same order (with the same comments) down below where the capture occurs. The test opens with the normal mock sycl namespace and mock USM functions, and, while labelled, maybe I could make it clearer to scroll down to get into the actual tests. Or maybe I should move those mocks to an #included file.

The test itself is second-order, which is why I've not commented on it before.  But the individual things being tested and cases being tested should be broken up into simple, self explanatory sections.

There are two correct solutions to this that I can see: 
1- prohibit pointers entirely.  Codeplay did this for exactly this reason.
2- Warn on nearly EVERY SINGLE pointer capture, whether wrapped in a type or not. Then have the library provide a type to expose pointers safely via the USM method.  I'd contend this is really what the users want as well.  (this type would be the one that would not warn^^).

> 
> ( * ) There's an old joke about this, goes something like:
> Q: How long will it take you to solve the halting problem?
> A: Let me do some calculations and get back to you....

erichkeane commented:
Now I see this, I thought of a third: Add a message to a common static-analysis tool to attempt to catch this.  This has the advantage of being allowed to be slower (meaning better analysis), work cross-TU (meaning better analysis), and  being less reliable (meaning more forgiving analysis).

cperkinsintel commented:
Do we have such a static analysis tool, or Clang plugin?  I can move this over there, or at least propose doing so.

erichkeane commented:
> Do we have such a static analysis tool, or Clang plugin? I can move this over there, or at least propose doing so.

I  don't believe we do.  There IS a clang static analyzer, but no one from Intel has been contributing (let alone for SYCL) as far as I know.

bader commented:
+1 for improving static analyzer, which seems to be the tool for any diagnostics requiring cross-TU analysis and it's more tolerant to "not 100% accurate" diagnostics.
https://reviews.llvm.org/D72362#1817682

I think some other existing checks might be worth moving to the static analyzer.
